R/make-traceability-matrix.R

```diff
@@ -332,38 +354,58 @@ get_exports <- function(pkg_source_path){
 #'
 #' @keywords internal
 filter_symbol_functions <- function(funcs){
-  ignore_functions <- c("\\%>\\%", "\\$", "\\[\\[", "\\[", "\\+")
-  pattern <- paste0("^(", paste(ignore_functions, collapse = "|"), ")$")
+  ignore_patterns <- c("\\%>\\%", "\\$", "\\[\\[", "\\[", "\\+", "\\%", "<-")
+  pattern <- paste0("(", paste(ignore_patterns, collapse = "|"), ")")
   funcs_return <- grep(pattern, funcs, value = TRUE, invert = TRUE)
   return(funcs_return)
 }
 
-#' list all functions defined in the package code
+#' list all top-level objects defined in the package code
 #'
-#' @inheritParams make_traceability_matrix
+#' This is primarily for getting all _functions_, but it also returns top-level
+#' declarations, regardless of type. This is intentional, because we also want
+#' to capture any global variables or anything else that could be potentially
+#' exported by the package.
+#'
+#' @inheritParams map_functions_to_scripts
 #'
 #' @return A data.frame with the columns `func` and `code_file` with a row for
-#'   every function defined in the package.
+#'   every top-level object defined in the package.
 #'
 #' @keywords internal
-get_all_functions <- function(pkg_source_path){
-
-  # Get all defined functions (following syntax: func <- function(arg), or setGeneric("func"))
-  r_files <- list.files(file.path(pkg_source_path, "R"), full.names = TRUE)
+get_toplevel_assignments <- function(pkg_source_path){
+  r_files <- list.files(
+    file.path(pkg_source_path, "R"),
+    full.names = TRUE, pattern = "\\.[Rr]$", recursive = TRUE
+  )
   pkg_functions <- purrr::map_dfr(r_files, function(r_file_i) {
-    file_text <- readLines(r_file_i) %>% suppressWarnings()
-    pattern <- paste0("^\\s*([[:alnum:]_\\.]+)\\s*(<\\-|=)\\s*function\\s*.*", "|",
-                      "^\\s*setGeneric\\s*\\(\\s*[\"|']([[:alnum:]_\\.]+)[\"|'].*")
-    function_calls <- file_text[grepl(pattern, file_text)]
-    function_names <- gsub(pattern, "\\1\\3", function_calls)
-    # TODO: these patterns ^ are still missing some things... need to address in future commit
+    exprs <- tryCatch(parse(r_file_i), error = identity)
+    if (inherits(exprs, "error")) {
+      warning("Failed to parse ", r_file_i, ": ", conditionMessage(exprs))
+      return(tibble::tibble(func = character(), code_file = character()))
+    }
```

Comment:
Do we want this warning to be triggered at every occurrence, or would it make sense to only do this if `verbose` is `TRUE`?

Reply:
`parse` errors should be very rare in packages that are being score (as it means an invalid R file).  I don't think there's a need to guard with verbose, but if I'm overlooking something, I guess the worst case is just that the MPN build is noisy and we come back to adjust then.

Sound okay to you?

Reply:
Yeah im fine with that logic. In my testing it didnt seem to come up as often as I expected, but was still triggered during the test (with actual _intentional_ code errors), so it's fine to keep as is.

Reply:
>  (with actual _intentional_ code errors)

Right, exactly.  On the MPN side: something's not getting through the initial build if it triggers a parse error, as far as I know.  And on our package development side: we're not releasing a package with a parse error.